linera-chain/src/certificate.rs

```diff
@@ -0,0 +1,304 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    borrow::Cow,
+    fmt::{Debug, Formatter},
+};
+
+use linera_base::{
+    crypto::{CryptoHash, Signature},
+    data_types::Round,
+    identifiers::BlobId,
+};
+use linera_execution::committee::{Committee, ValidatorName};
+use serde::{ser::Serializer, Deserialize, Deserializer, Serialize};
+
+use crate::{
+    block::{ConfirmedBlock, ValidatedBlock},
+    data_types::{
+        Certificate, CertificateValue, ExecutedBlock, HashedCertificateValue, LiteCertificate,
+        LiteValue,
+    },
+    ChainError,
+};
+
+/// Certificate for a [`ValidatedBlock`]` instance.
+pub type ValidatedBlockCertificate = CertificateT<ValidatedBlock>;
+
+/// Certificate for a [`ConfirmedBlock`] instance.
+pub type ConfirmedBlockCertificate = CertificateT<ConfirmedBlock>;
+
+/// Generic type representing a certificate for `value` of type `T`.
+pub struct CertificateT<T> {
```

Comment:
`GenericCertificate` ?

Reply:
I cannot call it `Certificate` b/c of the name clash with the _other_ `Certificate` type. I can call it `GenericCertificate` but it's longer and I'm not sure it's that usefull really. Almost everywhere we refer to a type alias (like `ConfirmedBlockCertificate`) and when we'll be ready to use `CertificateT<_>` in places where right now we use `Certificate` - then I'll simply remove the "old" type and rename this to `Certificate`.

Reply:
Done in https://github.com/linera-io/linera-protocol/pull/2803/commits/44ac99fe8f260336836c42560fae345227c8f3f7